sending messages: changed the group PM avatar

thisisnitish opened the pull request:
Fixes: #4870
As @alya mentioned:
At present, the group PM avatar shows first initials of the participants
, which doesn't look great and doesn't actually make it easy to figure
out at a glance who is in the conversation.
Instead of initials, we should use a simple "group" icon, similar to the
huddle icon we show in the left sidebar on desktop.

I've removed the `GroupAvatar` and added `FontAwesome` group icon.
Now, It will only show the group icon for group PM.
![Simulator Screen Shot - iPhone 11 - 2021-07-24 at 00 22 25](https://user-images.githubusercontent.com/48220369/126828582-95fb7f7c-5385-4d0b-b4f7-f60319bd7129.png)

thisisnitish commented:
@alya, Please have a look at this

alya commented:
Thanks @thisisnitish !

1. Can you please add screenshots in both dark and light mode?
2. I think a dark grey might look better in light mode than black -- can you try that and post a screenshot as well?

It also looks like you need to fix some checks.

thisisnitish commented:
Sure @alya I'll change the color of the icon from black to dark grey. If the icon will be in black color then it will not work in dark mode I guess, so it is better to change the color.
Yeah, I checked now, I've to fix some checks.
I will try to fix these issue as soon as possible. Thanks 🙂

thisisnitish commented:
Hey @alya below I've changed the colour of group icon from black to dark grey, I took screen shots in both light and dark mode. After seeing this, I think dark grey is fine for the light mode but what about dark mode, I think it will vey hard to see, isn't it?
![Simulator Screen Shot - iPhone 11 - 2021-07-27 at 00 36 05](https://user-images.githubusercontent.com/48220369/127044620-0264289d-55b4-49b2-9f58-7d49a3ac5f3c.png)
![Simulator Screen Shot - iPhone 11 - 2021-07-27 at 00 35 06](https://user-images.githubusercontent.com/48220369/127044637-0184104d-3381-4ee5-a28a-fdf7a7796d6f.png)

thisisnitish commented:
Also, @alya, I have changed the colour from dark grey to light grey, please give feedback for this 🙂. I think this colour will be fine for both the dark as well as light mode.
![Simulator Screen Shot - iPhone 11 - 2021-07-27 at 00 43 57](https://user-images.githubusercontent.com/48220369/127045637-feee88a9-de5b-4a74-9143-e6a5c2b43284.png)
![Simulator Screen Shot - iPhone 11 - 2021-07-27 at 00 44 03](https://user-images.githubusercontent.com/48220369/127045652-d48b1249-3cc7-4582-9253-35c2032039da.png)

alya commented:
Thanks for the screenshots! We should probably have a dark grey for light mode and a light grey for dark mode; I don't think we need to try to use the same color for both.

thisisnitish commented:
Sure @alya That could also be fine. I'll try to fix this issue as soon as possible 🙂

thisisnitish commented:
> Thanks for the screenshots! We should probably have a dark grey for light mode and a light grey for dark mode; I don't think we need to try to use the same color for both.

Hey @alya, I've fixed the issue according to your suggestions. Please have a look 🙂.

chrisbobbe commented:
Thanks, @thisisnitish!

Did I miss some discussion about this point that @alya raised in the issue 🙂 :

> To keep the groups visually distinct, we should continue to use random background colors for the group avatar.

Especially if we're still thinking of doing that (and probably even if not), you might find it easier to work within the `GroupAvatar` component, instead of replacing it with something else. For example, it already sets `borderRadius`, which would be helpful if we're putting the icon on a colored square with rounded corners. (Working within `GroupAvatar` would also have the benefit that we'd get the new style for free in other places we might start using `GroupAvatar` in the future.)

thisisnitish commented:
> Thanks, @thisisnitish!
> 
> Did I miss some discussion about this point that @alya raised in the issue 🙂 :
> 
> > To keep the groups visually distinct, we should continue to use random background colors for the group avatar.
> 
> Especially if we're still thinking of doing that (and probably even if not), you might find it easier to work within the `GroupAvatar` component, instead of replacing it with something else. For example, it already sets up a square with rounded corners, and it shows how to get the right background color for the recipients, by calling `colorHashFromString`.
> 
> Working within `GroupAvatar` would also have the benefit that we'd get the new style for free in other places we might start using `GroupAvatar` in the future.

Thanks @chrisbobbe for the suggestions. I also thought the same in very beginning 😁but then decided to go with the issue.
Yeah, offcourse, using `GroupAvatar` component would be the good idea. Let me try this 😊👍🏻.

chrisbobbe commented:
> Let me try this 😊👍🏻

Great, thanks! 🙂

I don't yet understand, though: did we decide to do the colored background (with unique colors for each set of recipients), as Alya mentioned on the issue, or no? It looks like we lose the colored background in the current revision of the PR.

@alya, do you have a preference? 🙂

Doing the colored background would affect the choice of icon color, I think. Maybe we'd just stick with `foregroundColorFromBackground` in that case (i.e., choose the icon's color based on the background color, in the same way that we currently choose the color for the initials)?

alya commented:
Ah, heh, I forgot that part when I was reviewing earlier -- sorry!

It would be great to see screenshots with a colored background; that's
probably the best approach.

On Tue, Jul 27, 2021 at 12:24 PM Chris Bobbe ***@***.***>
wrote:

> Let me try this 😊👍🏻
>
> Great, thanks! 🙂
>
> I don't yet understand, though: did we decide to do the colored background
> (with unique colors for each set of recipients), as Alya mentioned on the
> issue, or no? It looks like we lose the colored background in the current
> revision of the PR.
>
> @alya <https://github.com/alya>, do you have a preference? 🙂
>
> Doing the colored background would affect the choice of icon color, I
> think. Maybe we'd just stick with foregroundColorFromBackground in that
> case (i.e., choose the icon's color based on the background color, in the
> same way that we currently choose the color for the initials)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zulip/zulip-mobile/pull/4922#issuecomment-887773238>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAP6IUU6TT224FP3AASS3B3TZ4BWTANCNFSM5A4QQ65Q>
> .
>

thisisnitish commented:
Hey @alya and @chrisbobbe, I've used the `GroupAvatar` component and as @chrisbobbe said use the background color for group avatar inorder to visually distinct the groups. Also, I'm using `foregroundColorFromBackground` for changing the icon colour according to the background colour. Please have a look 🙂👍🏻
![Simulator Screen Shot - iPhone 11 - 2021-07-28 at 22 54 55](https://user-images.githubusercontent.com/48220369/127371435-86d53c96-8cdf-413f-ac36-c9ad3401ff82.png)
![Simulator Screen Shot - iPhone 11 - 2021-07-28 at 22 55 04](https://user-images.githubusercontent.com/48220369/127371446-9b5c9d47-15c7-43fd-851e-731a53abcebb.png)

alya commented:
Hm, I think it might look better with light gray on the dark backgrounds / dark gray on the light backgrounds. The white/black is a little stark. Could you try that and post screenshots again? Thanks!

thisisnitish commented:
> Hm, I think it might look better with light gray on the dark backgrounds / dark gray on the light backgrounds. The white/black is a little stark. Could you try that and post screenshots again? Thanks!

It would be helpeful if @chrisbobbe also see this. Also @alya this looks really good. Moreover, there could be a possibility where we will be using `GroupAvatar` component in future as @chrisbobbe said. Also, with this approach, the groups are visually distinct and people can figure out very easily and group icon colour will be depend on the background colour. Thanks 🙂.

chrisbobbe commented:
I think Alya is asking for screenshots after just a small change:
- Keep the colored backgrounds.
- But for the icons, use dark gray and light gray, instead of black and white.

I wonder if we should also try something with opacity? For example, to make a darker yellow icon on a lighter yellow background. Just a thought.

thisisnitish commented:
Hey @alya , @chrisbobbe I tried light grey on dark mode and dark grey on light mode. Please give the feedback. Should we stick with the previous one or should we go with this?
![Simulator Screen Shot - iPhone 11 - 2021-07-29 at 22 48 15](https://user-images.githubusercontent.com/48220369/127537611-94c609df-7010-4728-a4b8-2dc8a0b6df7d.png)
![Simulator Screen Shot - iPhone 11 - 2021-07-29 at 22 47 45](https://user-images.githubusercontent.com/48220369/127537618-4d4f3819-90b4-4f4a-a080-a0954ab4573e.png)

thisisnitish commented:
> Ah, sorry if the feedback was unclear -- since you added backgrounds, the color of the icon should be controlled by the *background* color at this point, not the app mode. So you'd still use foregroundColorFromBackground to set the color, but the options would be light gray vs. dark gray, not white vs. black.
> […](#)
> On Thu, Jul 29, 2021 at 10:30 AM Nitish Kumar ***@***.***> wrote: Hey @alya <https://github.com/alya> , @chrisbobbe <https://github.com/chrisbobbe> I tried light grey on dark mode and dark grey on light mode. Please give the feedback. Should we stick with the previous one or should we go with this? [image: Simulator Screen Shot - iPhone 11 - 2021-07-29 at 22 48 15] <https://user-images.githubusercontent.com/48220369/127537611-94c609df-7010-4728-a4b8-2dc8a0b6df7d.png> [image: Simulator Screen Shot - iPhone 11 - 2021-07-29 at 22 47 45] <https://user-images.githubusercontent.com/48220369/127537618-4d4f3819-90b4-4f4a-a080-a0954ab4573e.png> — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#4922 (comment)](https://github.com/zulip/zulip-mobile/pull/4922#issuecomment-889329550)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAP6IUWZXBMC2QDNM4A3SADT2GFZXANCNFSM5A4QQ65Q> .

Yes, But for the grey colours I didn't used `foregroundColorFromBackground` I just hard coded the colours so that, I can show you. Yes we can change from black and white to light grey and dark grey in `foregroundColorFromBackground`, so that It will change the icon colour according to the background.

thisisnitish commented:
Hey @alya, @chrisbobbe I've change the `foregroundColorFromBackground`. Now it will only show light grey and dark grey based on the background colour. Please have a look at this. Also I've not committed anything as of now, If you agree I'll commit it and push the changes 🙂👍🏻
![Simulator Screen Shot - iPhone 11 - 2021-07-30 at 23 03 37](https://user-images.githubusercontent.com/48220369/127691544-b71efbc9-8c59-4385-92a7-493369bd9574.png)
![Simulator Screen Shot - iPhone 11 - 2021-07-30 at 23 03 46](https://user-images.githubusercontent.com/48220369/127691552-3e070d5f-19b4-4ca3-a94f-68c3135fa0db.png)

thisisnitish commented:
Hey @alya , @chrisbobbe, as I said, I've updated the `foregroundColorFromBackground`. Above, I showed you dark and light grey on hard coded value, I made it just to show case you. Now, I've updated the `foregroundColorFromBackground` function and now the group icon will have light grey and dark grey colour based on the background colour dynamically. Thanks 🙂👍🏻
![Simulator Screen Shot - iPhone 11 - 2021-08-04 at 23 51 57](https://user-images.githubusercontent.com/48220369/128236454-cbd0ca78-6e9b-403c-a911-0e58ee27c142.png)
![Simulator Screen Shot - iPhone 11 - 2021-08-04 at 23 52 04](https://user-images.githubusercontent.com/48220369/128236476-6dd82551-addc-4ca3-97d7-1e8113295cf7.png)

chrisbobbe commented:
Thanks, @thisisnitish!

Would you mind trying an approach like the one I mentioned [above](https://github.com/zulip/zulip-mobile/pull/4922#issuecomment-888548905), just to see? I think it might look better than the gray, but it's hard to tell without trying it out. 🙂 

> I wonder if we should also try something with opacity or tinting? For example, to make a darker yellow icon on a lighter yellow background. Just a thought.

I'm imagining using the ['color'](https://www.npmjs.com/package/color) library, to make a new color for the icon, based on the unique, colored background color for the conversation item, maybe with `.fade()`, or maybe `.lighten()` or `.darken()`. You can see an example of using that library in 5bc0af1a6.

thisisnitish commented:
Sure @chrisbobbe I will give it a shot and tell you 🙂👍🏻.

thisisnitish commented:
Hey @chrisbobbe, as you suggested to try out few things, so I tried, Please have a look at this 🙂👍🏻.
![Simulator Screen Shot - iPhone 11 - 2021-08-05 at 23 38 56](https://user-images.githubusercontent.com/48220369/128400070-12e87d48-8c34-4d2d-a1d6-635d9b20ad25.png)
![Simulator Screen Shot - iPhone 11 - 2021-08-05 at 23 38 23](https://user-images.githubusercontent.com/48220369/128400079-2ae431fc-5e95-4ecd-8900-dee08726f69b.png)

alya commented:
I like how the colored ones look!

thisisnitish commented:
@alya me either, I want to see @chrisbobbe's response

chrisbobbe commented:
> This looks more beautiful 😅

I agree, I like it!

When you push this revision to GitHub, please [squash](https://zulip.readthedocs.io/en/latest/git/fixing-commits.html) the behavior change into a single commit, to keep the branch nice and tidy, and then force-push. We want to discard the commits that were implementing the older ideas, and just keep what's necessary for the new idea. 🙂 If you need to refactor some code (reorganize it without changing behavior), you can do so with some ["NFC" commits](https://github.com/zulip/zulip-mobile/blob/master/docs/glossary.md#nfc-or-nfc) before the interesting commit.

There's a doc [here](https://zulip.readthedocs.io/en/latest/git/using.html#examine-and-tidy-your-commit-history) that should help: see the section "Examine and tidy your commit history", and please follow the links there to read about Zulip's commit discipline and style guide. You can see as many examples of mergeable commits by viewing the project's history; Greg's ["secret" to using `git log -p`](https://github.com/zulip/zulip-mobile/blob/master/docs/howto/git.md#2-the-secret-to-using-git-log--p) will be really helpful for that. 🙂 

Thanks for your help!

thisisnitish commented:
Thanks @chrisbobbe and @alya 🥺for guiding me. Yeah I will learn about squash and I'll check the Zulip commit discipline and style guide. I'll do all these things as soon as possible. Once again thanks @chrisbobbe and @alya, looking forward to contribute more to Zulip community 😊

chrisbobbe commented:
> looking forward to contribute more to Zulip community 😊

Welcome! Please feel free to ask specific questions here or at https://chat.zulip.org/#narrow/stream/48-mobile (there might be easier than GitHub for more detailed discussions) 🙂

thisisnitish commented:
Hey @chrisbobbe I've used the git squash and merged all commits into the single commit. Basically, followed every single step as you suggested but here I'm seeing merge conflicts how to resolve that. Thanks

chrisbobbe commented:
Thanks for the revision! I've gone ahead and resolved that merge conflict, and force-pushed the result to your PR branch. 

To set your local branch to that result, please run `git checkout group_icon`. Then, if `me` is the alias you use for your own remote (it doesn't have to be, but if it's not, then adapt the command I've printed below), run

```
git pull --ff-only me group_icon
```

If you run `tools/test`, you'll see that it's reporting some errors, and I'll hand those back to you to fix, please. 🙂

Here are some hints I can give based on the `tools/test` output:

- Search for other places where we put the line `import Color from 'color'`. There's a line we've been putting just above that line, to tell Flow to ignore the fact that we don't have type definitions (yet) for that library.
- We'll want to remove the changes to `foregroundColorFromBackgroundColor`, I think, and just have it return 'black' or 'white'.

thisisnitish commented:
Sorry @chrisbobbe a very silly doubt, here `me` is getting an error, I guess you are saying in place of `me` I should've `origin`? Because my remote is `origin` and Zulip's is `upstream`. Please, correct me if I'm wrong? 🙂

chrisbobbe commented on src/common/GroupAvatar.js:
```diff
@@ -53,17 +56,17 @@ export default class GroupAvatar extends PureComponent<Props> {
       <Touchable onPress={onPress}>
         <View style={[styles.frame, frameSize]}>
-          <Text style={[styles.text, textSize]}>{initialsForGroupIcon(names)}</Text>
+          <FontAwesome size={38} name="group" color={iconColor} />
```
How did you choose 38 for the size? 🙂

Let's have the value be based on `GroupAvatar`'s `props.size`. Right now we're only using `GroupAvatar` in one place, and we're passing 48 there. But later, we might want to use it in other places, at other sizes, and it should still look appropriate when we do.

How about `props.size * 0.75`? That way, vertically at least, it can't possibly be at risk of running outside the rounded corners of the background.

chrisbobbe replied:
Also, instead of using `FontAwesome` here, directly, let's add another item in src/common/Icons.js, and import and use that.

thisisnitish replied:
I chose to use 38 because it looked appropriate to me, sure will change that as well as per your suggestions

thisisnitish replied:
Chris: I checked on `props.size * 0.75`, it is looking fine 😊👍🏻

chrisbobbe commented on src/common/GroupAvatar.js:
```diff
@@ -1,11 +1,14 @@
+import FontAwesome from 'react-native-vector-icons/FontAwesome';
 
+// $FlowFixMe[untyped-import]
+import Color from 'color';
```
Nit: This should go above the blank line. The blank line is meant to separate third-party imports (of code that isn't ours) from imports of our own code.

thisisnitish replied:
Sure, will do that 👍🏻

chrisbobbe commented on src/common/GroupAvatar.js:
```diff
@@ -53,17 +56,17 @@ export default class GroupAvatar extends PureComponent<Props> {
     return (
       <Touchable onPress={onPress}>
         <View style={[styles.frame, frameSize]}>
-          <Text style={[styles.text, textSize]}>{initialsForGroupIcon(names)}</Text>
```
`styles.text`, defined near the top of the file, isn't needed anymore and should be removed in this commit.

chrisbobbe reviewed (changes requested):
Thanks, @thisisnitish, this is getting close! See a few comments below.

chrisbobbe commented:
Thanks for the revision, @thisisnitish! Please [squash](https://zulip.readthedocs.io/en/latest/git/fixing-commits.html#squashing-commits) the two commits into one commit, for basically the reasons I mentioned at https://github.com/zulip/zulip-mobile/pull/4922#issuecomment-894290104. 🙂

thisisnitish commented:
@chrisbobbe, I've squashed the commits as you said, Please have a look 😊.

chrisbobbe commented on src/common/GroupAvatar.js:
```diff
@@ -53,17 +53,17 @@ export default class GroupAvatar extends PureComponent<Props> {
     return (
       <Touchable onPress={onPress}>
         <View style={[styles.frame, frameSize]}>
-          <Text style={[styles.text, textSize]}>{initialsForGroupIcon(names)}</Text>
```
Ah, sorry I missed this until now: it looks like this was the only place where we were using the `initialsForGroupIcon` function, so it's now dead code. So let's remove it, and also remove src/common/__tests__/initialsForGroupIcon-test.js. Those removals should both happen in the same commit as each other; could be squashed with the existing commit or could be a new commit after it.

thisisnitish replied:
Sure, will remove it, even I also missed `initialsForGroupIcon` function 😅. I will do both removal of the `initialsForGroupIcon` function and src/common/tests/initialsForGroupIcon-test.js in a single commit and then I will squash both the commits into a single commit. Squashing would be fine because, this is a not a very huge change and also single commit would be easy to understand 🙂.

chrisbobbe reviewed (changes requested):
Thanks, @thisisnitish! See a comment below about something I didn't catch until now (sorry!).

Also, for the commit message:

```
sending messages: added group icon in group PM conversations

As discussed:
- Created a `IconGroup` icon in `Icons.js` so that it will be used as
group PM avatar in group PM conversation

- Color of the icon will be darker as compare to it's background, but both
of them will have same color, only the icon will have darker tone as
compare to it's background

- Used `colorHashFromString` function with `Color` library for icon color,
so that icon will have the same color as it's background but with the darker
tone
```

For the summary line, let's use something like

```
GroupAvatar: Use "group" icon, instead of initials, for group PMs.
```

It's a bit more specific about the "before" and "after" of the change, it uses a more accurate/precise prefix (here, and in many cases, the name of the relevant file works well), and it uses the [imperative mood](https://en.wikipedia.org/wiki/Imperative_mood).

At the end of the commit-message body, please be sure to say `Fixes: #4870`, as you've done in the PR description. It needs to be in both places. 🙂

Currently, the body is basically a summary of the code changes made in the commit. This kind of summary unnecessary here, where the changes are easy to follow with `git log -p`. 😉 It would be more valuable for the body to say _why_ we're making these changes to `GroupAvatar`. That's basically done by the `Fixes: #4870` line; Alya's description in #4870 is very clear about the problem to be solved.

So I think this would be enough:

```
GroupAvatar: Use "group" icon, instead of initials, for group PMs.

Fixes: #4870
```

Or, if you want, you could briefly summarize #4870 in the body, above the `Fixes:` line. But I don't think that's highly necessary here.
